packages/common/src/tools/TaskQueue.ts

```diff
@@ -1,16 +1,27 @@
+import assert from 'assert'
+
 import { Logger } from './Logger'
 
 export class TaskQueue<T> {
   private queue: T[] = []
-  private isExecuting = false
 
   constructor(
     private executeTask: (task: T) => Promise<void>,
     private logger: Logger,
-  ) {}
+    private workers = 1,
+  ) {
+    assert(
+      workers > 0 && Number.isInteger(workers),
+      'workers needs to be a positive integer',
+    )
+  }
+
+  private get allWorkersBusy() {
+    return this.workers === 0
+  }
 
   addIfEmpty(task: T) {
-    if (this.queue.length === 0 && !this.isExecuting) {
+    if (this.queue.length === 0 && !this.allWorkersBusy) {
```

Comment:
This has now changed, where if there are multiple workers, but only some are busy the queue is considered empty. The correct logic should be that the queue is only empty if there are no queued items and there are no items being executed right now.

Reply:
Good catch, changed.